### DISCUSSION TITLE
Handling error

### DISCUSSION THREAD

**Discussion by survuvi** [modular/modular#234] [Q&A]
I would like to handle error with a message if the value passed to the function is not a number, but I get this error:

```python

fn your_function(x: F32) -> StringRef:

    var result: StringRef = "value"
    
    if x == 0:
        return zero()
    
    elif x > 0 and x < 9 or x < 0:
        result = "small"
        return result

    else:
        result = "big"
        return result
    
fn zero() -> StringRef:
    
    let result: StringRef = "zero"

    return result

try:
    print(your_function(a))
except:
    print("Something went wrong")


```

_error: Expression [16]:38:29: use of unknown declaration 'a'
        print(your_function(a))_

**Comment by mfranzon** [comment]
I think that the problem is because that this piece of code is part of a main (which is the process running under the notebook) so, the undefined variable error is raised by the master process not from the actual code you are running.
In fact, you can easily test this behavior with this simple piece of code:
```python
def foo(a) -> object:
    return (a + 3.45) < [1, 2, 3]

a = 1 #  if commented Expression error

try:
    foo(a)
except:
    print("exception")
```
If you run this, the try/except works fine. If you run the function `foo` outside the try/except it raises a `TypeError`, if you do not assigned the variable, the master raises the `Expression` error

**Reply by survuvi** [reply]
I tried also raise but without success:

```python

@raises
fn your_function(x: F32) -> StringRef:
    
    if not isinstance(x, (F32)):
        raise TypeError("Argument must be a number")

    var result: StringRef = "value"
    
    if x == 0:
        return zero()
    
    elif x > 0 and x < 9 or x < 0:
        result = "small"
        return result

    else:
        result = "big"
        return result
    
fn zero() -> StringRef:   
    let result: StringRef = "zero"
    return result

print(your_function(3)) #output should be: small

print(your_function("a")) #output should be: Argument must be a number

```

_error: Expression [6]:7:12: use of unknown declaration 'isinstance', 'fn' declarations require explicit variable declarations
    if not isinstance(x, (F32)):

error: Expression [6]:41:24: invalid call to 'your_function': argument #0 cannot be converted from 'StringLiteral' to 'SIMD[f32, 1]'
    print(your_function("a")) #output should be: Argument must be a number

Expression [6]:5:1: function declared here
fn your_function(x: F32) -> StringRef:_

**Comment by survuvi** [comment]
Now it works but with def, not works with fn

```python

def foo(a) -> object:    
    if a == 0:
        print("zero")        
    elif a > 0 and a < 9 or a < 0:
        print("small")
    elif a >= 9:
        print("big")
try:
    a = "k"
    foo(a)
except:
    print("Something went wrong")

```

**Reply by mfranzon** [reply]
Yep, from my understanding `def` raise an exception error like in python. The `fn` functions has not the same behaviour but you can define an `fn` function that raises, as written in the [changelog](https://docs.modular.com/mojo/changelog.html#week-of-2023-04-17)

Example:
```python
fn raise_an_error() raises:
    raise Error("I'm an error!")
try:
   raise_an_error()
except:
    print("exception")
```

**Reply by maevadevs** [reply]
I ended up here after working on the HelloMojo notebook. Error Handling is not defined there except for one paragraph

> Both [`def` and `fn`] support raising exceptions, but this must be explicitly declared on a `fn` with the `raises` function effect, placed after the function argument list.

I think it would be helpful to add @mfranzon 's example [above](https://github.com/modularml/mojo/discussions/234#discussioncomment-5982954) on the notebook for reference. This makes more sense.

Making it more general:

```py
fn raise_an_error(num: Int) raises -> Int:
    if num < 0:
        raise Error("I'm an error!")
    
    return num +1
    

try:
    print(raise_an_error(-1))
except:
    print("exception")
```
